google/cloud/security/common/gcp_type/firewall_rule.py

```diff
@@ -249,12 +276,27 @@ def applies_to_all(self):
         return self._applies_to_all
 
     @property
+    def any_value(self):
+        """Returns whether this rule matches any value.
+
+        Returns:
+          bool: Whether this rule matches any value.
+        """
+        if self._any_value is None:
+            self._any_value = False
+            if self.rules == '*':
+                self._any_value = True
+        return self._any_value
+
+    @property
     def expanded_rules(self):
         """Returns an expanded set of ports.
 
         Returns:
           list: A list of every string port number.
         """
+        if self.any_value:
+            return []
```

Comment:
Expanded rules is a dict in one path and a list in the other. I don't like the special casing to check any_value outside of the check for the protected attribute and with a different return path.

I recommend something like:

if not self._expanded_rules:
  self._expanded_rules = {}
  if not self.any_value:
    for rule in self.rules:
...
return self._expanded_rules

Also make sure the docstring matches the type (dict)

Reply:
Done